### ISSUE TITLE
Support fastboot

### ISSUE THREAD

**Issue by seawatts** [auth0-community/ember-simple-auth-auth0#50]

**Comment by sukima** [comment]
Fastboot fails because this addon imports auth0-lock which depends on a `document` global and fastboot does not offer this.

```console
ReferenceError: document is not defined
    at Object.200 (/Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:151240:11)
    at s (/Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:121228:255)
    at /Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:121228:306
    at Object.281.../atom/index (/Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:158009:14)
    at s (/Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:121228:255)
    at /Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:121228:306
    at Object.253.../store/index (/Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:155805:14)
    at s (/Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:121228:255)
    at /Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:121228:306
    at Object.241../gravatar/actions (/Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:154348:25)
    at s (/Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:121228:255)
    at /Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:121228:306
    at Object.<anonymous> (/Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:151305:14)
    at Object.201../index (/Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:151327:4)
    at s (/Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:121228:255)
    at e (/Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:121228:426)
    at /Users/suki/Private/source/gifting-stitches/tmp/static_boot_build-input_base_path-rE6tpuG2.tmp/fastboot/vendor.js:121228:444
    at ContextifyScript.Script.runInContext (vm.js:32:29)
    at VMSandbox.eval (/Users/suki/Private/source/gifting-stitches/node_modules/fastboot/src/vm-sandbox.js:14:14)
    at /Users/suki/Private/source/gifting-stitches/node_modules/fastboot/src/ember-app.js:142:15
    at Array.forEach (native)
    at EmberApp.loadAppFiles (/Users/suki/Private/source/gifting-stitches/node_modules/fastboot/src/ember-app.js:139:21)
    at EmberApp.retrieveSandboxedApp (/Users/suki/Private/source/gifting-stitches/node_modules/fastboot/src/ember-app.js:187:10)
    at new EmberApp (/Users/suki/Private/source/gifting-stitches/node_modules/fastboot/src/ember-app.js:50:21)
    at FastBoot._buildEmberApp (/Users/suki/Private/source/gifting-stitches/node_modules/fastboot/src/index.js:114:17)
    at new FastBoot (/Users/suki/Private/source/gifting-stitches/node_modules/fastboot/src/index.js:52:10)
    at StaticBootBuild.build (/Users/suki/Private/source/gifting-stitches/node_modules/ember-cli-staticboot/lib/broccoli/staticboot.js:29:15)
    at /Users/suki/Private/source/gifting-stitches/node_modules/broccoli-plugin/read_compat.js:93:34
    at tryCatch (/Users/suki/Private/source/gifting-stitches/node_modules/rsvp/dist/lib/rsvp/-internal.js:215:12)
    at invokeCallback (/Users/suki/Private/source/gifting-stitches/node_modules/rsvp/dist/lib/rsvp/-internal.js:230:13)
    at publish (/Users/suki/Private/source/gifting-stitches/node_modules/rsvp/dist/lib/rsvp/-internal.js:198:7)
    at flush (/Users/suki/Private/source/gifting-stitches/node_modules/rsvp/dist/lib/rsvp/asap.js:85:5)
    at _combinedTickCallback (internal/process/next_tick.js:73:7)
    at process._tickCallback (internal/process/next_tick.js:104:9)
```

**Comment by roschaefer** [comment]
I'm [migrating an existing ember app to fastboot](https://github.com/roschaefer/rundfunk-mitbestimmen/pull/360) and learned that we have two incompatible plugins, including this one. Therefore I would love to see this issue resolved.

I never made an ember addon  fastboot compatible. That's why I feel not quite confident to tackle this alone.

What's the status on this issue? How about pair programming with someone experienced?

**Comment by XaserAcheron** [comment]
There's a PR for this (#74) that fixes the document reference error, though it's a bit out of date since we've switched from `bower` to `npm` imports for the auth0 libs.

The main thing it does is run the libs through [fastboot-transform](https://github.com/kratiahuja/fastboot-transform), which looks pretty easy to use. Almost _too_ easy -- I'll play around with it once the current release stuff is sorted out.

Apologies for being a bit slow on things.

**Comment by roschaefer** [comment]
@XaserAcheron why apologize? I'm pretty happy that you take up maintaining this plugin.

OK, so I experimented a little:
1. Create a plain ember app from scratch
2. `ember install ember-cli-fastboot`
3. `ember install ember-simple-auth`
4. `ember install ember-simple-auth-auth0`
5. edit `package.json` and point this plugin on branch `#develop`
6. `rm -rf package-lock.json node_modules bower_components` and `npm install` - bower is gone
7. Create an index route and controller
8. Follow the README of this plugin (almost, see below)
9. Add cookie session [as described here](https://github.com/simplabs/ember-simple-auth#fastboot)

Up to this point, the ember app does not crash. If you disable javascript you are always unauthenticated and you see a `Login` button that does nothing. When you enable javascript the app gets loaded and you can open the Auth0 widget.

If you add the `ApplicationRouteMixin` from this plugin, you hit the `ReferenceError: document is not defined`.

I have a few questions now:
1. I never questioned the `ApplicationRouteMixin` what is the purpose?
2. We are always unauthenticated in ember-fastboot:
    1. What are we supposed to sent (cookie?) on the client?
    2. What are we supposed to do on the server?
3. How can we open the Auth0 Widget on the client in ember-fastboot?

Here's the example app: https://github.com/roschaefer/EmberSimpleAuthAuth0Fastboot
See the reference error in action on branch `reference_error`.

Can we pair-program (e.g. on hangouts) to clear up the confusion?
Shall we contact Mark on slack, the maintainer of ember-simple-auth0?

**Comment by roschaefer** [comment]
After reading [this comment](https://github.com/auth0-community/ember-simple-auth-auth0/pull/94#issuecomment-411765978) I googled `auth0 hosted login` and found [this amazing blog post](https://auth0.com/blog/build-and-authenticate-an-emberjs-3-application/) by @SarahJay55 and the [corresponding repository](https://github.com/SarahJay55/ember-js-auth). I forked the example app, added `ember-simple-auth` and `ember-fastboot` and refactored it. I'm happy with it, server-side-rendering seems to work. It ships minimal code, has no dependency except `auth0-js` and works with an ordinary authenticator instead of a rather complex route mixin. The only difference is the hosted login instead of using lock on the app itself.

Here's my fork: https://github.com/roschaefer/ember-js-auth
Would love to hear your feedback. I'll now try to make my real-world ember app fastboot compliant without using `ember-simple-auth-auth0` based on the code of the example app.

**Comment by sukima** [comment]
@roschaefer Last time I looked into this repo (about 1.5 years ago) I ran into the same problem. My solution was the same as yours. Drop the whole lock thing and just popup a hosted login. In fact I didn't even need `auth0-js` to do this. I used [Torii](https://vestorly.github.io/torii/) instead and auth0 only provided the hosted login and the JWT token which were available without the auth0 JavaScript code.

**Comment by roschaefer** [comment]
@sukima: Cool, could you share some code how to set it up correctly?

I'm running into an issue on the server now. Just after authentication, on entering the ember app, I see an error on the server: 
```
Error while processing route: authentication.callback Cannot read property 'hash' of undefined TypeError: Cannot read property 'hash' of undefined
    at WebAuth.parseHash (/home/robert/Development/rundfunk-mitbestimmen/frontend/tmp/broccoli_persistent_filterreplace_filter-output_path-soNhrANX.tmp/assets/node_modules/auth0-js/build/auth0.js:8277:1)
    at Promise (/home/robert/Development/rundfunk-mitbestimmen/frontend/tmp/broccoli_persistent_filterreplace_filter-output_path-soNhrANX.tmp/assets/frontend/authenticators/auth0.js:15:1)
```
I believe these two lines are the culprit: https://github.com/auth0/auth0.js/blob/ee463df4a03357bbd27ec139e47fb0f327214669/src/web-auth/index.js#L151
https://github.com/auth0/auth0.js/blob/ee463df4a03357bbd27ec139e47fb0f327214669/src/helper/window.js#L12

The client is able to authenticate though, which would make sense.

@sukima if I can get around `auth0-js` somehow, that would help.

**Comment by sukima** [comment]
@roschaefer In a moment of ill-advised spring cleaning it looks like I deleted the code on the project where I did this.

What I remember is that I installed ember-simple-auth and ember-simple-auth-token. I then installed torii and had e-s-a login page show a button. When the user clicked the button it used torii's authenticate method which pointed to auth0.com's HTML page. Torii did the hard wrk of showing a popup window, auth0 did the authentication. I set auth0 to use torii's default redirect URL and torii pulled the JWT token from that redirect and closed the popup window. Then ember-simple-auth took the JWT and stored it in localStorage and redirected the main app to the authenticated route.

Basically I followed torii's documentation to setup a 3rd party OAuth2 popup window. And auth0.com just hosted the HTML and its OAuth redirect URL pointed back to my app's redirect URL (which torii provided).

**Comment by XaserAcheron** [comment]
Bit of a random interjection, but I was doing a brief bit of research on this, and in HTTP-land, the URL hash is never actually sent to the server, by design, so there just isn't anything to parse.

Y'all continue on -- Auth0 recommends using their hosted login nowadays anyway -- but I'll make a fix in this addon soonish to check for `window` before calling parseHash. I'm not a fastboot user myself so I really don't know how far this will go, but at the very least the addon shouldn't crash. ;)

**Comment by roschaefer** [comment]
That makes sense. From [Wikipedia](https://en.wikipedia.org/wiki/Fragment_identifier):
```The fragment identifier functions differently to the rest of the URI: its processing is exclusively client-side with no participation from the web server, though the server typically helps to determine the MIME type, and the MIME type determines the processing of fragments. When an agent (such as a Web browser) requests a web resource from a Web server, the agent sends the URI to the server, but does not send the fragment. Instead, the agent waits for the server to send the resource, and then the agent processes the resource according to the document type and fragment value.```

So, I guess this is the way to go. The server does not parse the JWT. The client has to wait for the Ember app to be downloaded and parse the JWT on the client. All the server can do is to show some kind of login page: "Hang tight, while we're logging you in.. ".

**Comment by sukima** [comment]
Yes, user enters credentials in the auth0 hosted page. Per OAuth2 on success it redirects the user back to a page which can parse the token from the query string. In a server based system the server takes the token and then redirects the client to a page **without** the token in it.

In your case (client app) Torii handles the hard part by opening a popup window which shows the auth0 login page hosted at auth0.com when the user logs into auth0 it redirects the _popup window_ to your ember app's hosted domain with a special HTML served from Torii's public directory (included in the build) which saves the token from the query string to localStorage.

Your ember app behind the popup notices the localStorage change and closes the popup then passes the JWT token saved in the localStorage to ember-simple-auth where you can either pass it to your server for verification or parse in the client.
